cupy/core/carray.cuh

```diff
@@ -209,58 +229,82 @@ public:
     return (*const_cast<CArray<T, 0>*>(this))[idx];
   }
 
-  __device__ T& operator[](int i) {
+  __device__ T& operator[](const ptrdiff_t* idx) {
+    return *reinterpret_cast<T*>(data_);
+  }
+
+  __device__ T operator[](const ptrdiff_t* idx) const {
+    return (*const_cast<CArray<T, 0>*>(this))[idx];
+  }
+
+  __device__ T& operator[](ptrdiff_t i) {
     return *reinterpret_cast<T*>(data_);
   }
 
-  __device__ T operator[](int i) const {
+  __device__ T operator[](ptrdiff_t i) const {
     return (*const_cast<CArray<T, 0>*>(this))[i];
   }
 };
 
 template <int ndim>
 class CIndexer {
 private:
-  int size_;
-  int shape_[ndim];
-  int index_[ndim];
+  ptrdiff_t size_;
+  ptrdiff_t shape_[ndim];
+  ptrdiff_t index_[ndim];
 
 public:
-  __device__ int size() const {
+  __device__ ptrdiff_t size() const {
     return size_;
   }
 
-  __device__ void set(int i) {
-    unsigned int a = i;
-    for (int dim = ndim; --dim > 0; ) {
-      unsigned int s = shape_[dim];
-      index_[dim] = (a % s);
-      a /= s;
+  __device__ void set(ptrdiff_t i) {
+    if (ndim == 0) {
+      index_[0] = i;
+      return;
     }
-    if (ndim > 0) {
-      index_[0] = a;
+    if (size_ > 1LL << 31) {
+      size_t a = i;
+      for (int dim = ndim; --dim > 0; ) {
+        size_t s = shape_[dim];
+        index_[dim] = (a % s);
+        a /= s;
+      }
+      if (ndim > 0) {
```

Comment:
I think this condition is unnecessary because `ndim == 0` case is exluded by the condition at the beginning of this function.